assembly/src/lib.rs

```diff
@@ -43,6 +45,63 @@ const MODULE_PATH_DELIM: &str = "::";
 type ProcMap = BTreeMap<String, Procedure>;
 type ModuleMap = BTreeMap<String, ProcMap>;
 
+// MODULE PROVIDER
+// ================================================================================================
+
+/// The module provider is now a simplified version of a module cache. It is expected to evolve to
+/// a general solution for the module lookup
+///
+/// TODO compute a procedure index deterministically from a module path & procedure label. The
+/// initial expected layout for the index is `[u8; 20]`. The module provider should map procedures
+/// from these indexes, and this is expected to result in code simplification + optimization for
+/// the compiler.
+pub trait ModuleProvider {
+    /// Fetch source contents provided a module path
+    fn get_source(&self, path: &str) -> Option<&str>;
+}
+
+#[derive(Default)]
+struct ModuleProviderController {
+    // instead of creating a map to available source, we provide a lazy load approach where the
+    // user will query for a given module path, and we will iterater the available sources
+    providers: Vec<Box<dyn ModuleProvider>>,
+}
```

Comment:
What benefit do we get by adding the controller? i.e., could we define the assembler as:
```Rust
pub struct Assembler {
    module_provider: Box<dyn ModuleProvider>,
    ...
}
```
This way, it would be completely up to the `ModuleProvider` implementation to specify the logic for mapping indexes to modules. For example, if there were 2 or more libraries, a single map for all of them could be created at instantiation time, and then we won't have to iterate over all module providers.

Reply:
The idea is to create a very basic controller for now, but extends its functionality when the action of providing a module goes beyond fetching a source.

We will have some questions such as "should we resolve modules as `domain+name` or always full path?" that will be handled by this structure, maybe even with unit tests coverage.

We might want to delegate this problem to the implementors of `ModuleProvider` if we want to leave the procedure id resolution up to them. However, we might end with divergent module resolution that can potentially leak into the language.

wdyt?

Reply:
My tendency is to do something simpler now and then add functionality to it later as we need. The model we are moving towards now is that, at the assembly level, we'll map `[u8; 20] -> ModuleAst` (maybe we should call type `[u8; 20]` procedure ID or something like that). This is primarily because procedure names will be erased at the serialized AST level, and thus we won't be able to use procedure names, full paths etc.

Basically, I'm not seeing a need to support a more generic resolution methodology than `[u8; 20] -> `ModuleAst` (the current `&str -> &str` is just a temporary step). But of course I can be mistaken :)